gokart/gcs_obj_metadata_client.py

```diff
@@ -71,30 +75,63 @@ def add_task_state_labels(
             if update_response is None:
                 logger.error(f'failed to patch object {obj} in bucket {bucket} and object {obj}.')
 
+    @staticmethod
+    def _normalize_labels(labels: Optional[dict[str, Any]]) -> dict[str, str]:
+        return {str(key): str(value) for key, value in labels.items()} if labels else {}
+
     @staticmethod
     def _get_patched_obj_metadata(
         metadata: Any,
-        task_params: Optional[dict[Any, str]] = None,
+        task_params: Optional[dict[str, str]] = None,
+        custom_labels: Optional[dict[str, Any]] = None,
     ) -> Union[dict, Any]:
         # If metadata from response when getting bucket and object information is not dictionary,
         # something wrong might be happened, so return original metadata, no patched.
         if not isinstance(metadata, dict):
             logger.warning(f'metadata is not a dict: {metadata}, something wrong was happened when getting response when get bucket and object information.')
             return metadata
 
-        if not task_params:
+        if not task_params and not custom_labels:
             return metadata
         # Maximum size of metadata for each object is 8 KiB.
         # [Link]: https://cloud.google.com/storage/quotas#objects
-        max_gcs_metadata_size, total_metadata_size, labels = 8 * 1024, 0, []
-        for label_name, label_value in task_params.items():
+        normalized_task_params_labels = GCSObjectMetadataClient._normalize_labels(task_params)
+        normalized_custom_labels = GCSObjectMetadataClient._normalize_labels(custom_labels)
+        max_gcs_metadata_size, total_metadata_size = 8 * 1024, 0
+        # There is a possibility that the keys of user-provided labels(custom_labels) may conflict with those generated from task parameters (task_params_labels).
+        # However, users who utilize custom_labels are no longer expected to search using the labels generated from task parameters.
+        # Instead, users are expected to search using the labels they provided.
+        # Therefore, in the event of a key conflict, the value registered by the user-provided labels will take precedence.
+        total_metadata_size, labels, has_seen_keys = GCSObjectMetadataClient._add_labels_with_size_limitation(
+            normalized_custom_labels, total_metadata_size, max_gcs_metadata_size
+        )
+        _, labels, _ = GCSObjectMetadataClient._add_labels_with_size_limitation(
+            normalized_task_params_labels, total_metadata_size, max_gcs_metadata_size, labels, has_seen_keys
+        )
```

Comment:
I think these methods are a bit confusing. I think we can separate two phase.

1. merge `normalized_task_params_labels` and `normalized_custom_labels`
2. delete oversized labels 

I think you shouldn't multiple things in a method.